src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt

```diff
@@ -0,0 +1,78 @@
+package com.safehill.kcrypto
+
+import dev.turingcomplete.kotlinonetimepassword.HmacAlgorithm
+import dev.turingcomplete.kotlinonetimepassword.TimeBasedOneTimePasswordConfig
+import dev.turingcomplete.kotlinonetimepassword.TimeBasedOneTimePasswordGenerator
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.toJavaDuration
+
+/**
+ * @param digits number of digits in the otp.
+ * @param validDuration the validity of otp.
+ * The generated otp will be the same for half the time step if the same secret is used.
+ * Use [isValid] to check if the otp is valid within the validity duration.
+ */
+class SafehillOTP(
+    digits: Int,
+    validDuration: Duration
+) {
+    private val halfTimeStep = validDuration / 2
+
+    private val otpConfig = TimeBasedOneTimePasswordConfig(
+        codeDigits = digits,
+        hmacAlgorithm = HmacAlgorithm.SHA1,
+        timeStep = halfTimeStep.inWholeMicroseconds,
```

Comment:
Why did you pick microseconds and not milliseconds?

Reply:
I tried seconds for the first time and generated a code which was valid for 1 seconds.  In this case, I was generating otp which was valid for 1/2 = 0.5 seconds which was being rounded to 0 seconds and the tests were failing.
Then I switched to milliseconds and at the last moment I thought why not use microseconds and increase the precision and thus, it was just a moment of decision to use microseconds.
But upon reviewing the source code of TimeBasedOneTimePasswordConfig, it seems they are working with milliseconds as well.

Reply:
I would use milliseconds, yes